File: src/main/java/gargoyle/netsquares/prop/simple/IntegerProperty.java
package gargoyle.netsquares.prop.simple;

import gargoyle.netsquares.prop.RWProperty;

import java.io.Serializable;
import java.util.Objects;

public final class IntegerProperty extends RWProperty<Integer> implements Comparable<IntegerProperty>, Serializable {
    public IntegerProperty(Integer value) {
        super(value);
    }

    public IntegerProperty() {
        this(0);
    }

    public IntegerProperty(int value) {
        super(value);
    }

    @Override
    public int compareTo(IntegerProperty o) {
        return Integer.compare(get(), o.get());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RWProperty)) return false;
        if (!super.equals(o)) return false;
        RWProperty<?> that = (RWProperty<?>) o;
        return Objects.equals(get(), that.get());
    }

    public int getValue() {
        return get();
    }

    public void setValue(int value) {
        set(value);
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/simple/IntegerProperty.kt
package gargoyle.netsquares.prop.simple

import gargoyle.netsquares.prop.RWProperty

class IntegerProperty(value: Int = 0) : RWProperty<Int>(value), Comparable<IntegerProperty> {

    override fun compareTo(other: IntegerProperty): Int {
        return get().compareTo(other.get())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RWProperty<*>) return false
        if (!super.equals(other)) return false
        return get() == other.get()
    }

    var value: Int
        get() = get()
        set(value) {
            set(value)
        }

    override fun hashCode(): Int = get().hashCode()
}
